### PR TITLE
Fix run operation return codes (#1377)

### LINKED ISSUE #1377: Fail with nonzero exit code if run-operation task encounters an unhandled exception
### Issue description
If a run-operation macro fails with an exception (ie. unable to connect to database, invalid SQL, etc), dbt should exit with a nonzero exit code. Moreover, run-operations should generally go through a similar code-path to models, seeds, etc, such that this happens automatically.

### Results
The info logs showed that an exception had occurred, but the task exited with exit code = 0

### System information
The output of `dbt --version`:
```
0.13.0
```

### PR DESCRIPTION
**Pull request by beckjake**
Fixes #1377 

When we run an operation, set the return code on error.

Also, since wilt-chamberlain totally broke the run-operation command in multiple ways, I fixed all that. And added tests so I don't do that again.

I also added a commit call that happens after successfully executing the macro. Maybe that would be better as a flag? I just wanted it for testing, but I can't really see _not_ wanting to commit.

### CODE REVIEW

File: core/dbt/task/run_operation.py [hunk 1]
```diff
@@ -22,19 +21,44 @@ def _get_macro_parts(self):
     def _get_kwargs(self):
         return dbt.utils.parse_cli_vars(self.args.args)
 
-    def run(self):
+    def run_unsafe(self):
```
Comment (cmcarthur) [inline core/dbt/task/run_operation.py]: can you make this `_run_unsafe`

**Comment by drewbanin** [comment]
@cmcarthur yeah, I think Wilt is the right time to release this for real.

@beckjake I think it's actually pretty important that we _don't_ add a commit at the end of operations. Some queries (like vacuum / analyze on Redshift) are not allowed to run inside of transactions. I think that if people want to run transactional code through operations, they should add the `begin` and `commit` explicitly.

If we do this, then we need to be really sure that we don't begin a transaction implicitly, otherwise the query won't be committed. Do we have a good way of ensuring that connections never have open connections when the operations SQL runs?

File: core/dbt/task/run_operation.py [hunk 2]
```diff
@@ -22,19 +21,44 @@ def _get_macro_parts(self):
...
+            adapter.commit_if_has_connection()
 
         return res
+
+    def run(self):
+        try:
+            result = self._run_unsafe()
+        except dbt.exceptions.Exception as exc:
+            logger.error(
+                'Encountered a dbt exception while running a macro: {}'
```
Comment (drewbanin) [inline core/dbt/task/run_operation.py]: This is an exception that dbt is aware of, but it's not a dbt exception per se. I just don't want folks to think that dbt did something wrong if their operations fail. Can we make this say:
```
Encountered an error while running an operation:
```

**Comment by beckjake** [comment]
>Some queries (like vacuum / analyze on Redshift) are not allowed to run inside of transactions. I think that if people want to run transactional code through operations, they should add the begin and commit explicitly.

Ok, I can do that instead then.

>If we do this, then we need to be really sure that we don't begin a transaction implicitly, otherwise the query won't be committed.

I think that's already true. If your macro is implemented via `call statement` currently like I kind of assume it is, you must make sure not set `auto_begin=False`. Right?


>Do we have a good way of ensuring that connections never have open connections when the operations SQL runs?

I assume you mean "... connections never have open transactions ...". Yeah. `adapter.clear_transaction()` is reliable, in my experience.

File: core/dbt/task/run_operation.py [hunk 3]
```diff
@@ -22,19 +21,43 @@ def _get_macro_parts(self):
...
+    def run(self):
+        try:
+            result = self._run_unsafe()
+        except dbt.exceptions.Exception as exc:
+            logger.error(
+                'Encountered an error while running operation: {}'
+                .format(exc)
+            )
+            logger.debug('', exc_info=True)
+            return False, None
```
[the remark this answers is not in the document]
Comment (beckjake) [reply core/dbt/task/run_operation.py]: Sure, though in the RPC case I'd probably reach for overriding the `run` method entirely (the message would end up in the response logs anyway)

**Pull request merged**